doc/source/notebooks/tailor/external-mean-function.pct.py

```diff
@@ -66,32 +75,66 @@ def generate_data(num_functions=10, N=1000):
 # %%
 _ = plt.plot(Xs, F)
 
-
 # %% [markdown]
 # We generate the meta and test tasks.
 
+
 # %%
-def generate_meta_and_test_tasks(num_datapoints, num_meta, num_test):
-    N = 1000
+def generate_meta_and_test_tasks(num_datapoints, num_meta=1000, num_test=200, N=50):
+    """Generates meta-task datasets {D_i} = {x_i, y_i} and target task training
+    and test data {\tilde{x}, \tilde{y}} (Fortuin and Rätsch, 2019).
+
+    Args:
+        num_datapoints: The number of training points, \tilde{n} in Table S1.
+        num_meta: The number of meta-tasks ("sampled functions").
+        num_test: The number of test tasks ("unseen functions").
+        N: Number of sampled data points per function.
+
+    Returns:
+        A tuple (meta, test) where
+        - meta: List of num_meta pairs of arrays (X, Y) of size (n, 1) each.
+        - test: List of num_test pairs of pairs of arrays of sizes
+                (((num_datapoints, 1), (num_datapoints, 1)), 
+                 ((N - num_datapoints, 1), (N - num_datapoints, 1))).
+    """
     Xs, F = generate_data(num_functions=num_meta + num_test, N=N)
-    meta_indices = [np.random.permutation(N)[:num_datapoints] for _ in range(num_meta)]
-    test_indices = [np.random.permutation(N)[:num_datapoints] for _ in range(num_test)]
     meta = []
-    for i, mi in enumerate(meta_indices):
-        Y = F[mi, i][:, None] + 1e-1 * np.random.randn(num_datapoints, 1)
-        meta.append((Xs[mi], Y))
+    sd = 1e-1  # Standard deviation for normally-distributed observation noise.
+    for i in range(num_meta):
+        # We always use all data points of the curve to train the mean function,
+        # i.e. n_i = N.
+        noise = sd * np.random.randn(N, 1)
+        Y = F[:, i][:, None] + noise
+        meta.append((Xs, Y))
     test = []
-    for i, ti in enumerate(test_indices):
-        Y = F[ti, num_meta + i][:, None] + 1e-1 * np.random.randn(num_datapoints, 1)
-        test.append(((Xs[ti], Y), (Xs, F[:, num_meta + i][:, None])))
+    rand_indices = np.random.permutation(N)
+    train_i = sorted(rand_indices[:num_datapoints])
+    test_i = sorted(rand_indices[num_datapoints:])
+    for i in range(num_test):
+        # Form target training set, \tilde{D}_{train} (see Figure 1).
+        noise = sd * np.random.randn(num_datapoints, 1)
+        train_X = Xs[train_i]
+        train_Y = F[train_i, num_meta + i][:, None] + noise
+        # Form target test set, \tilde{D}_{test}, which is disjoint from the
+        # target traininig set as it is in the original implementation.
+        test_X = Xs[test_i]
+        test_Y = F[test_i, num_meta + i][:, None]
+        # Form target tasks' datasets \tilde{D} as a pair of pairs,
+        # (\tilde{X}, \tilde{Y}) and (\tilde{X}*, \tilde{Y}*).
+        test.append(((train_X, train_Y), (test_X, test_Y)))
```

Comment:
Could you briefly explain what you changed here? Did you just clean up *how* things are generated (but the returned values will be the same as before, if the random seed was the same), or does this change the logic of *what* is generated? (To make it easier to review your changes!)

Reply:
Thanks for the question! 
1. Previously each meta-training episode was done on only `num_datapoints` points, but it should be all `N` points (so no need for any `meta_indices` anymore where `num_datapoints` points are picked randomly).
2. The target datasets (called `test` for meta-testing) are split into train and test (`train_X, train_Y, test_X, test_Y`) . Previously, the train was a subset of the test, but after clarifying it they are actually disjoint.

So the logic has changed a little bit, although they are probably not that important. Hope this clears things up, let me know if you see more areas for improvement!